[FLINK-8518][Table SQL / API] Add support for extract Epoch

snuyanzin opened the pull request:
## What is the purpose of the change
This is the final part of the ticket https://issues.apache.org/jira/browse/FLINK-8518
which implements support of `EXTRACT` of `EPOCH` e.g.
```sql
SELECT EXTRACT(EPOCH FROM current_date),
       EXTRACT(EPOCH FROM current_timestamp);
```
 (`EXTRACT` of `ISODOW`, `ISOYEAR`, `DECADE` was implemented within https://github.com/apache/flink/pull/17422)

## Verifying this change

The tests are added at 
_flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala_,
_flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/TemporalTypesTest.scala_,
_flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/validation/ScalarFunctionsValidationTest.scala_

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: ( no )
  - The runtime per-record code paths (performance sensitive): ( no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: ( no)
  - The S3 file system connector: ( no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

snuyanzin commented:
@flinkbot run azure

snuyanzin commented:
@flinkbot run azure

snuyanzin commented:
@flinkbot run azure

snuyanzin commented:
@slinkydeveloper , @twalthr 
sorry for the poke
could you please have a look once you have time

slinkydeveloper commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/DateTimeUtils.java:
```diff
@@ -1137,25 +1137,29 @@ private static long convertExtract(
             case EPOCH:
-                // TODO support it
-                throw new TableException("EPOCH is unsupported now.");
+                if (type instanceof TimestampType || type instanceof DateType) {
```
Can you use `type.isAnyOf` here?

snuyanzin replied:
yes, done

slinkydeveloper commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/DateTimeUtils.java:
```diff
@@ -1137,25 +1137,29 @@ private static long convertExtract(
+                    throw new TableException(startUnit + " for " + type + " is unsupported now.");
+                }
             case MICROSECOND:
                 if (type instanceof TimestampType) {
```
`type.is`

snuyanzin replied:
done

slinkydeveloper reviewed (commented):
Just 2 reccomendations here:

* We have another validation layer for Table, located in `BuiltInFunctionDefinitions`, which I'm touching right now, exactly for this function :) https://github.com/apache/flink/pull/19190 Depending on who merges first, we need to reflect changes there as well. I suggest we merge this one first.
* Can you add the new test cases to the new test infra using `BuiltInFunctionsITCase`?

snuyanzin commented:
@slinkydeveloper thank you for your feedback
I've rebased the PR to be able to use `org.apache.flink.table.planner.functions.ExtractFunctionITCase#getTestSetSpecs`

Also added `org.apache.flink.table.planner.functions.ExtractFunctionITCase`

slinkydeveloper commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/ExtractFunctionITCase.java:
```diff
@@ -0,0 +1,94 @@
+                                TIMESTAMP().nullable(),
+                                TIMESTAMP().nullable(),
+                                TIMESTAMP().nullable())
+                        .testSqlResult(
```
Why are you testing only sql? Shouldn't this be supported by table api as well?

snuyanzin replied:
It looks not for every timeunit table api is supported
I will have a look if it is possible to support others

slinkydeveloper replied:
> I will have a look if it is possible to support others

Thank you, We should align the two as much as we can, and in this particular case it seems as easy as just adding the missing enum variants

snuyanzin replied:
I added support for all except `DOW`, `ISODOW`, `DOY`, `ISOYEAR`.
For those 4 I have a question: from one side `extract` is supported for `TimeIntervalUnit` i.e. need to add these 4 elements there, from the other side these 4 elements are not interval units... Does it make sense add extract for `TimePointUnit` e.g. or leave it like it is?

slinkydeveloper replied:
What Calcite is doing for those?

twalthr commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/typeutils/SymbolUtil.java:
```diff
@@ -184,6 +184,27 @@
+                TimeUnitRange.MILLENNIUM,
+                "TIME_UNIT_RANGE",
+                "MILLENNIUM");
+
```
remove empty line here

twalthr commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/ExtractFunctionITCase.java:
```diff
@@ -0,0 +1,284 @@
+        return Stream.of(
+                TestSetSpec.forFunction(BuiltInFunctionDefinitions.EXTRACT)
+                        .onFieldsWithData(
+                                LocalDateTime.of(2000, 1, 31, 11, 22, 33, 123456789),
```
Can you add a test for extracting `EPOCH` on `DATE`? As far as I see it in the code, we also support it now?

twalthr commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/ExtractFunctionITCase.java:
```diff
@@ -0,0 +1,284 @@
+                TestSetSpec.forFunction(BuiltInFunctionDefinitions.EXTRACT)
+                        .onFieldsWithData(
+                                LocalDateTime.of(2000, 1, 31, 11, 22, 33, 123456789),
+                                LocalDateTime.of(2020, 2, 29, 1, 56, 59, 987654321),
```
What is the current behavior of `EXTRACT` on `TIMESTAMP_LTZ`? Can we add a test for it? At least for testing the error behavior?

twalthr commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala:
```diff
@@ -1801,6 +1801,12 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
 
   @Test
   def testExtract(): Unit = {
```
We don't need to add tests to the old test base. Can we remove this method entirely because `ExtractFunctionITCase` is much nicer.

twalthr commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/ExtractFunctionITCase.java:
```diff
@@ -0,0 +1,284 @@
+                                null,
+                                BIGINT().nullable())
+                        .testResult(
+                                call("EXTRACT", TimeIntervalUnit.MILLENNIUM, $("f0")),
```
use `$("f0").extract(...)` for Table API everywhere, instead of lookup via name

twalthr commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/validation/ScalarFunctionsValidationTest.scala:
```diff
@@ -168,6 +168,12 @@ class ScalarFunctionsValidationTest extends ScalarTypesTestBase {
   }
 
+  @Test
+  def testEpochWithTimeWhichIsUnsupported(): Unit = {
```
Let's not add new tests to outdated test bases. Use `ExtractFunctionITCase`.

twalthr commented on docs/content/docs/dev/table/functions/systemFunctions.md:
```diff
@@ -146,6 +146,7 @@ For Table API, please use `_` for spaces (e.g., `DAY_TO_HOUR`).
 |                          | `MICROSECOND`                  |
 | `DOY` _(SQL-only)_       |                                |
 | `DOW` _(SQL-only)_       |                                |
+| `EPOCH` _(SQL-only)_     |                                |
```
You updated `TimeIntervalUnit.java`. So many symbols in this table are not `SQL-only` anymore?
